drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,417 @@
+	I2C_CC32XX_IDLE_MODE = 0,
+	/* I2C is currently performing a write operation */
+	I2C_CC32XX_WRITE_MODE,
+	/* I2C is currently performing a read operation */
+	I2C_CC32XX_READ_MODE,
+	/* I2C error has occurred */
+	I2C_CC32XX_ERROR = 0xFF
+};
+
+struct i2c_cc32xx_config {
+	u32_t base;
+	u32_t bitrate;
+	unsigned int irq_no;
+};
+
+struct i2c_cc32xx_data {
+	struct k_sem mutex;
+	struct k_sem transfer_complete;
+
+	volatile enum i2c_cc32xx_state state;
+
+	struct i2c_msg msg; /* Cache msg for transfer state machine */
+	u16_t  slave_addr; /* Cache slave address for ISR use */
+};
+
+static void configure_i2c_irq(const struct i2c_cc32xx_config *config);
+
+static int i2c_cc32xx_configure(struct device *dev, u32_t dev_config_raw)
+{
+	u32_t base = DEV_BASE(dev);
+	u32_t bitrate_id;
+
+	if (!(dev_config_raw & I2C_MODE_MASTER)) {
+		return -EINVAL;
+	}
+
+	if (dev_config_raw & I2C_ADDR_10_BITS) {
+		return -EINVAL;
+	}
+
+	switch (I2C_SPEED_GET(dev_config_raw)) {
+	case I2C_SPEED_STANDARD:
+		bitrate_id = 0;
+		break;
+	case I2C_SPEED_FAST:
+		bitrate_id = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	MAP_I2CMasterInitExpClk(base, CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC,
+				bitrate_id);
+
+	return 0;
+}
+
+static void _i2c_cc32xx_prime_transfer(struct device *dev, struct i2c_msg *msg,
+				      u16_t addr)
+{
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t base = DEV_BASE(dev);
+
+	/* Initialize internal counters and buf pointers: */
+	data->msg = *msg;
+	data->slave_addr = addr;
+
+	/* Start transfer in Transmit mode */
+	if (IS_I2C_MSG_WRITE(data->msg.flags)) {
+
+		/* Specify the I2C slave address */
+		MAP_I2CMasterSlaveAddrSet(base, addr, false);
+
+		/* Update the I2C state */
+		data->state = I2C_CC32XX_WRITE_MODE;
+
+		/* Write data contents into data register */
+		MAP_I2CMasterDataPut(base, *((data->msg.buf)++));
+
+		/* Start the I2C transfer in master transmit mode */
+		MAP_I2CMasterControl(base, I2C_MASTER_CMD_BURST_SEND_START);
+
+	} else {
+		/* Start transfer in Receive mode */
+		/* Specify the I2C slave address */
+		MAP_I2CMasterSlaveAddrSet(base, addr, true);
+
+		/* Update the I2C mode */
+		data->state = I2C_CC32XX_READ_MODE;
+
+		if (data->msg.len < 2) {
+			/* Start the I2C transfer in master receive mode */
+			MAP_I2CMasterControl(base,
+				       I2C_MASTER_CMD_BURST_RECEIVE_START_NACK);
+		} else {
+			/* Start the I2C transfer in burst receive mode */
+			MAP_I2CMasterControl(base,
+					    I2C_MASTER_CMD_BURST_RECEIVE_START);
+		}
+	}
+}
+
+static int i2c_cc32xx_transfer(struct device *dev, struct i2c_msg *msgs,
+			       u8_t num_msgs, u16_t addr)
+{
+	struct i2c_cc32xx_data *data = DEV_DATA(dev);
+	u32_t base = DEV_BASE(dev);
+	int retval = 0;
+
+	/* Acquire the driver mutex */
+	k_sem_take(&data->mutex, K_FOREVER);
+
+	/* Iterate over all the messages */
+	for (int i = 0; i < num_msgs; i++) {
+
+		/* Protect prime_transfer from irq: */
+		MAP_I2CMasterIntDisable(base);
```

Comment:
If we are here it means that the last transfer has finished and the interrupts should have already been disabled. If they are not the ISR routine will attempt to handle them even if the driver itself isn't expecting it. We should not need this line. This is just a suggestion to double check the concept. I may be wrong.

Reply:
Yes, ideally, shouldn't be needed.  Removed the protection.